### PR TITLE
Add `async_task::spawn_local` variant that includes caller in panics

### PR DESCRIPTION
**Pull request by mgsloan**
For debugging #21020.  Copy-modified [from async_task here](https://github.com/smol-rs/async-task/blob/ca9dbe1db9c422fd765847fa91306e30a6bb58a9/src/runnable.rs#L432)

Release Notes:

- N/A

### CODE REVIEW

File: crates/gpui/src/executor.rs [hunk 1]
```diff
@@ -456,6 +463,69 @@ impl ForegroundExecutor {
...
+                self.location
+            );
+            unsafe { self.map_unchecked_mut(|c| &mut *c.inner).poll(cx) }
+        }
+    }
+
+    // Wrap the future into one that checks which thread it's on.
+    let future = Checked {
+        id: thread_id(),
+        inner: ManuallyDrop::new(future),
```
Comment (mikayla-maki) [inline crates/gpui/src/executor.rs:522]: What's going on with this manually drop?

Comment (mgsloan) [reply crates/gpui/src/executor.rs:525]: I'm not sure, this is copy modified [from async_task here](https://github.com/smol-rs/async-task/blob/ca9dbe1db9c422fd765847fa91306e30a6bb58a9/src/runnable.rs#L432)

Comment (mgsloan) [reply crates/gpui/src/executor.rs:527]: Added this github permalink to the code

Comment (mikayla-maki) [reply crates/gpui/src/executor.rs:527]: Oh cool, maybe it's too control drop order or something.

**Pull request merged**